base/src/main/java/proguard/optimize/inline/BaseLambdaInliner.java

```diff
@@ -0,0 +1,357 @@
+        // calls to the old function to calls to this new function.
+        inlinedLambdaMethod = methodWithoutLambdaParameter;
+    }
+
+    @Override
+    public void visitConstantInstruction(Clazz consumingClass, Method consumingMethod, CodeAttribute consumingMethodCodeAttribute, int consumingMethodCallOffset, ConstantInstruction constantInstruction) {
+        consumingClass.constantPoolEntryAccept(constantInstruction.constantIndex, this);
+        Method invokeInterfaceMethod = referencedInterfaceConstant.referencedMethod;
+        Clazz invokeInterfaceClass = referencedInterfaceConstant.referencedClass;
+
+        // Check if this is an invokeinterface call to the lambda class invoke method
+        if (invokeInterfaceClass.getName().equals(interfaceClass.getName()) &&
+                invokeInterfaceMethod.getName(invokeInterfaceClass).equals(lambdaInvokeMethod.getName(lambdaClass)) &&
+                invokeInterfaceMethod.getDescriptor(invokeInterfaceClass).equals(bridgeDescriptor)
+        ) {
+            partialEvaluator.visitCodeAttribute(consumingClass, consumingMethod, consumingMethodCodeAttribute);
+            TracedStack tracedStack = partialEvaluator.getStackBefore(consumingMethodCallOffset);
+
+            // If we have a lambda with n arguments we have to skip those n arguments and then get the instance of the lambda used for calling invokeinterface.
+            int lambdaInstanceStackIndex = ClassUtil.internalMethodParameterCount(invokeInterfaceMethod.getDescriptor(invokeInterfaceClass));
+
+            consumingMethodCodeAttribute.instructionAccept(
+                consumingClass,
+                consumingMethod,
+                tracedStack.getTopActualProducerValue(lambdaInstanceStackIndex).instructionOffsetValue().instructionOffset(0),
+                new LambdaInvokeUsageReplacer(consumingMethodCallOffset)
+            );
+        }
+    }
+
+    @Override
+    public void visitInterfaceMethodrefConstant(Clazz clazz, InterfaceMethodrefConstant interfaceMethodrefConstant) {
+        referencedInterfaceConstant = interfaceMethodrefConstant;
+    }
+
+    private class LambdaInvokeUsageReplacer implements InstructionVisitor {
+        private final int possibleLambdaInvokeCallOffset;
+
+        public LambdaInvokeUsageReplacer(int possibleLambdaInvokeCallOffset) {
+            this.possibleLambdaInvokeCallOffset = possibleLambdaInvokeCallOffset;
+        }
+
+        @Override
+        public void visitVariableInstruction(Clazz clazz, Method method, CodeAttribute codeAttribute, int offset, VariableInstruction variableInstruction) {
+            // Uses same codeAttributeEditor as the casting remover
+            Instruction tracedInstruction = SourceTracer.traceParameter(partialEvaluator, codeAttribute, offset);
+            if (tracedInstruction.canonicalOpcode() == Instruction.OP_ALOAD) {
+                variableInstruction = (VariableInstruction) tracedInstruction;
```

Comment:
What if the tracedInstruction is not a `VariableInstruction`? Then we should throw an exception?